src/main/scala/org/broadinstitute/dsde/firecloud/service/TrialService.scala

```diff
@@ -71,110 +72,128 @@ final class TrialService
     case x => throw new FireCloudException("unrecognized message: " + x.toString)
   }
 
-  private def enableUsers(managerInfo: UserInfo, userEmails: Seq[String]): Future[PerRequestMessage] = {
-    // TODO: Handle multiple users
-    require(userEmails.size == 1, "For the time being, we can only enable one user at a time.")
-
-    val userEmail = userEmails.head
+  private def buildEnableUserStatus(userInfo: WorkbenchUserInfo, currentStatus: Option[UserTrialStatus]): UserTrialStatus = {
+    val needsProject = currentStatus match {
+      case None => true
+      case Some(statusObj) => statusObj.state match {
+        case Some(Disabled) => true
+        case _ => false // either an invalid transition or noop
+      }
+    }
+    if (needsProject) {
+      val trialProject = trialDAO.claimProjectRecord(WorkbenchUserInfo(userInfo.userSubjectId, userInfo.userEmail))
+      UserTrialStatus(userId = userInfo.userSubjectId, state = Some(Enabled), enabledDate = Instant.now, billingProjectName = Some(trialProject.name.value))
+    } else {
+      currentStatus.get.copy(state = Some(Enabled))
+    }
+  }
 
-    // Define what to overwrite in user's status
-    val statusTransition: UserTrialStatus => UserTrialStatus =
-      status => status.copy(state = Some(Enabled), enabledDate = Instant.now)
+  private def enableUserPostProcessing(updateStatus: Attempt, prevStatus: Option[UserTrialStatus], newStatus: UserTrialStatus): Unit = {
+    if (updateStatus != StatusUpdate.Success && newStatus.billingProjectName.isDefined) {
+      logger.info(
+        s"The user '${newStatus.userId}' failed to be enabled, releasing the billing project '${newStatus.billingProjectName.get}' back into the available pool.")
+      trialDAO.releaseProjectRecord(RawlsBillingProjectName(newStatus.billingProjectName.get))
+    }
+  }
 
-    updateUserState(managerInfo, userEmail, statusTransition)
+  private def enableUsers(managerInfo: UserInfo, userEmails: Seq[String]): Future[PerRequestMessage] = {
+    executeStateTransitions(managerInfo, userEmails, buildEnableUserStatus, enableUserPostProcessing)
   }
 
-  private def disableUsers(managerInfo: UserInfo, userEmails: Seq[String]): Future[PerRequestMessage] = {
-    // TODO: Handle multiple users
-    require(userEmails.size == 1, "For the time being, we can only disable one user at a time.")
+  private def buildDisableUserStatus(userInfo: WorkbenchUserInfo, currentStatus: Option[UserTrialStatus]): UserTrialStatus = {
+    currentStatus.get.copy(state = Some(Disabled), billingProjectName = None)
+  }
 
-    val userEmail = userEmails.head
+  private def disableUserPostProcessing(updateStatus: Attempt, prevStatus: Option[UserTrialStatus], newStatus: UserTrialStatus): Unit = {
```

Comment:
The `Unit` return type feels wrong to me. It seems like these transition steps should return at least true/false at a minimum to let the caller know that it did something or failed to do something. 

Same feedback for the other transition function `enableUserPostProcessing`.

Reply:
the post processing isn't really doing the state transition just some persistence based on the state transition state. i'm not sure what i would do with an output parameter...